core/trino-main/src/main/java/io/trino/server/security/oauth2/OidcDiscoveryModule.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.security.oauth2;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.inject.Binder;
+import com.google.inject.Provides;
+import com.nimbusds.oauth2.sdk.GeneralException;
+import com.nimbusds.oauth2.sdk.id.Issuer;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderConfigurationRequest;
+import com.nimbusds.openid.connect.sdk.op.OIDCProviderMetadata;
+import io.airlift.configuration.AbstractConfigurationAwareModule;
+import io.airlift.http.client.HttpStatus;
+import io.airlift.http.client.StringResponseHandler;
+import io.airlift.json.ObjectMapperProvider;
+
+import javax.inject.Singleton;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static java.lang.String.format;
+
+class OidcDiscoveryModule
+        extends AbstractConfigurationAwareModule
+{
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
+
+    @Override
+    protected void setup(Binder binder) {}
+
+    @Provides
+    @Singleton
+    public OAuth2Endpoints createOAuth2Endpoints(OAuth2Config config, NimbusHttpClient httpClient)
+    {
+        try {
+            return fetchOidcMetadata(config, httpClient);
+        }
+        catch (URISyntaxException | JsonProcessingException | GeneralException | IllegalStateException e) {
+            throw new RuntimeException("Error creating config using OpenID Provider issuer discovery mechanism: " + e.getMessage(), e);
+        }
+    }
+
+    private OAuth2Endpoints fetchOidcMetadata(OAuth2Config config, NimbusHttpClient httpClient)
```

Comment:
I don't think we should do http calls during guice setup. Let's have a dedicated load method like `io.trino.server.security.HeaderAuthenticatorManager#loadHeaderAuthenticator`